### PR TITLE
Handle invalid CSS selectors in menus and commands

### LINKED ISSUE #30: Uncaught SyntaxError: Failed to execute 'webkitMatchesSelector' on 'Element': '0' is not a valid selector.
[Enter steps to reproduce below:]

1. ...
2. ...

**Atom Version**: 0.179.0
**System**: Mac OS X 10.9.3
**Thrown From**: [command-palette](https://github.com/atom/command-palette) package, v0.34.0

### Stack Trace

Uncaught SyntaxError: Failed to execute 'webkitMatchesSelector' on 'Element': '0' is not a valid selector.

```
At /Applications/Atom.app/Contents/Resources/app/src/command-registry.js:117

Error: Failed to execute 'webkitMatchesSelector' on 'Element': '0' is not a valid selector.
  at Error (native)
  at CommandRegistry.module.exports.CommandRegistry.findCommands (/Applications/Atom.app/Contents/Resources/app/src/command-registry.js:117:121)
  at CommandPaletteView.module.exports.CommandPaletteView.show (/Applications/Atom.app/Contents/Resources/app/node_modules/command-palette/lib/command-palette-view.js:70:32)
  at CommandPaletteView.module.exports.CommandPaletteView.toggle (/Applications/Atom.app/Contents/Resources/app/node_modules/command-palette/lib/command-palette-view.js:49:21)
  at atom-workspace.<anonymous> (/Applications/Atom.app/Contents/Resources/app/node_modules/command-palette/lib/command-palette-view.js:21:21)
  at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/Applications/Atom.app/Contents/Resources/app/src/command-registry.js:243:29)
  at /Applications/Atom.app/Contents/Resources/app/src/command-registry.js:3:61
  at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:558:16)
  at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:396:22)
  at HTMLDocument.module.exports.WindowEventHandler.onKeydown (/Applications/Atom.app/Contents/Resources/app/src/window-event-handler.js:172:20)

```

### Commands

```
  4x -4:00.6 core:undo (atom-text-editor.editor.is-focused)
     -0:00.4 command-palette:toggle (atom-text-editor.editor.is-focused)
```

### Config

```json
{
  "core": {
    "excludeVcsIgnoredPaths": false,
    "disabledPackages": [
      "sublime-tabs",
      "linter-coffeelint"
    ],
    "audioBeep": false
  }
}
```

### Installed Packages

```coffee
# User
coffee-lint, v0.7.3
color-picker, v1.4.4
file-icons, v1.4.8
highlight-selected, v0.8.0
jsformat, v0.7.18
jshint, v1.2.0
language-marko, v0.4.8
linter, v0.10.0
pain-split, v1.2.0
project-manager, v1.14.1

# Dev
No dev packages
```

### LINKED ISSUE #32: Uncaught SyntaxError: Failed to execute 'webkitMatchesSelector' on 'Element': 'coffee-lint:lint-current-file' is not a valid selector.
[Enter steps to reproduce below:]

1. press shift-cmd-p 

**Atom Version**: 0.184.0
**System**: Mac OS X 10.10.1
**Thrown From**: [command-palette](https://github.com/atom/command-palette) package, v0.34.0

### Stack Trace

Uncaught SyntaxError: Failed to execute 'webkitMatchesSelector' on 'Element': 'coffee-lint:lint-current-file' is not a valid selector.

```
At /Applications/Atom.app/Contents/Resources/app/src/command-registry.js:117

Error: Failed to execute 'webkitMatchesSelector' on 'Element': 'coffee-lint:lint-current-file' is not a valid selector.
  at Error (native)
  at CommandRegistry.module.exports.CommandRegistry.findCommands (/Applications/Atom.app/Contents/Resources/app/src/command-registry.js:117:121)
  at CommandPaletteView.module.exports.CommandPaletteView.show (/Applications/Atom.app/Contents/Resources/app/node_modules/command-palette/lib/command-palette-view.js:70:32)
  at CommandPaletteView.module.exports.CommandPaletteView.toggle (/Applications/Atom.app/Contents/Resources/app/node_modules/command-palette/lib/command-palette-view.js:49:21)
  at atom-workspace.<anonymous> (/Applications/Atom.app/Contents/Resources/app/node_modules/command-palette/lib/command-palette-view.js:21:21)
  at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/Applications/Atom.app/Contents/Resources/app/src/command-registry.js:243:29)
  at /Applications/Atom.app/Contents/Resources/app/src/command-registry.js:3:61
  at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:558:16)
  at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:396:22)
  at HTMLDocument.module.exports.WindowEventHandler.onKeydown (/Applications/Atom.app/Contents/Resources/app/src/window-event-handler.js:182:20)

```

### Commands

```
     -9:03.2 fuzzy-finder:toggle-file-finder (atom-text-editor.editor)
  2x -9:00.6 core:backspace (atom-text-editor.editor.mini)
     -8:57.8 editor:consolidate-selections (atom-text-editor.editor.mini)
     -8:57.8 core:cancel (atom-text-editor.editor.mini)
     -8:50.8 fuzzy-finder:toggle-file-finder (atom-text-editor.editor)
     -8:47.9 editor:consolidate-selections (atom-text-editor.editor.mini)
     -8:47.9 core:cancel (atom-text-editor.editor.mini)
     -3:40.3 fuzzy-finder:toggle-file-finder (atom-text-editor.editor)
     -3:38.2 editor:consolidate-selections (atom-text-editor.editor.mini)
     -3:38.2 core:cancel (atom-text-editor.editor.mini)
     -3:37.3 fuzzy-finder:toggle-file-finder (atom-text-editor.editor)
     -3:35.7 editor:consolidate-selections (atom-text-editor.editor.mini)
     -3:35.7 core:cancel (atom-text-editor.editor.mini)
  9x -3:15.6 core:backspace (atom-text-editor.editor.mini)
  3x -0:15.6 command-palette:toggle (atom-text-editor.editor.mini)
```

### Config

```json
{
  "core": {
    "followSymlinks": true,
    "disabledPackages": [
      "composer",
      "behat-atom",
      "ruby-block-converter"
    ]
  }
}
```

### Installed Packages

```coffee
# User
atom-lint, v0.20.1
atom-runner, v2.3.0
autocomplete-plus, v2.3.0
coffee-lint, v0.7.3
git-log, v0.3.0
git-tab-status, v1.5.3
jsonlint, v1.0.1
language-twig, v1.5.2
linter, v0.12.0
linter-csslint, v0.0.11
linter-flow, v0.1.3
linter-haml, v0.1.1
linter-php, v0.0.11
linter-rubocop, v0.2.2
linter-ruby, v0.1.4
linter-scss-lint, v0.0.14
markdown-format, v2.5.0
php-twig, v3.0.0
phpunit, v1.0.9
phpunit-snippets, v0.1.0
preview-plus, v1.1.19
project-manager, v1.15.5
rsense, v0.6.0
save-session, v0.11.5
script, v2.18.0
slash-closer, v0.7.1
symbols-tree-view, v0.6.1
symfony-snippets, v0.3.0
sync-settings, v0.2.1

# Dev
No dev packages
```

### LINKED ISSUE #5098: On Windows 7, if you right-click on a file, then a stack trace is returned
1. In Atom, open a file
2. Right-click on the file
3. Then, wait for below error to appear

**Atom Version**: 0.174.0
**System**: Microsoft Windows 7 Home Premium 
**Thrown From**: Atom Core

### Stack Trace

Uncaught SyntaxError: Failed to execute 'webkitMatchesSelector' on 'Element': '`atom-workspace atom-text-editor:not(.mini)' is not a valid selector.

```
At c:\Users\Wilson\AppData\Local\atom\app-0.174.0\resources\app\src\context-menu-manager.js:95

Error: Failed to execute 'webkitMatchesSelector' on 'Element': '`atom-workspace atom-text-editor:not(.mini)' is not a valid selector.
  at Error (native)
  at c:\Users\Wilson\AppData\Local\atom\app-0.174.0\resources\app\src\context-menu-manager.js:95:32
  at Array.filter (native)
  at ContextMenuManager.module.exports.ContextMenuManager.templateForEvent (c:\Users\Wilson\AppData\Local\atom\app-0.174.0\resources\app\src\context-menu-manager.js:94:42)
  at ContextMenuManager.module.exports.ContextMenuManager.showForEvent (c:\Users\Wilson\AppData\Local\atom\app-0.174.0\resources\app\src\context-menu-manager.js:165:27)
  at HTMLDocument.<anonymous> (c:\Users\Wilson\AppData\Local\atom\app-0.174.0\resources\app\src\window-event-handler.js:141:33)
  at HTMLDocument.handler (c:\Users\Wilson\AppData\Local\atom\app-0.174.0\resources\app\src\space-pen-extensions.js:112:34)
  at HTMLDocument.jQuery.event.dispatch (c:\Users\Wilson\AppData\Local\atom\app-0.174.0\resources\app\node_modules\space-pen\vendor\jquery.js:4681:9)
  at HTMLDocument.elemData.handle (c:\Users\Wilson\AppData\Local\atom\app-0.174.0\resources\app\node_modules\space-pen\vendor\jquery.js:4359:46)

```

### Commands

```
  2x -0:00.0 tree-view:remove (ol.tree-view.full-menu.list-tree.has-collapsable-children.focusable-panel)
```

### Config

```json
{
  "core": {
    "themes": [
      "atom-dark-ui",
      "atom-dark-syntax"
    ],
    "audioBeep": false
  },
  "editor": {
    "fontSize": 14,
    "autoIndent": false,
    "autoIndentOnPaste": false,
    "tabLength": 4,
    "softWrap": true,
    "confirmCheckoutHeadRevision": false,
    "invisibles": {}
  }
}
```

### Installed Packages

```coffee
# User
autocomplete-paths, v0.9.1
autocomplete-plus, v1.1.0
go-plus, v3.0.13

# Dev
No dev packages
```

### LINKED ISSUE #5536: Uncaught SyntaxError: Failed to execute 'webkitMatchesSelector' on 'Element': '0' is not a valid selector.
[Enter steps to reproduce below:]

1. ...
2. ...

**Atom Version**: 0.179.0
**System**: Mac OS X 10.9.3
**Thrown From**: Atom Core

### Stack Trace

Uncaught SyntaxError: Failed to execute 'webkitMatchesSelector' on 'Element': '0' is not a valid selector.

```
At /Applications/Atom.app/Contents/Resources/app/src/package.js:708

Error: Failed to execute 'webkitMatchesSelector' on 'Element': '0' is not a valid selector.
  at Error (native)
  at /Applications/Atom.app/Contents/Resources/app/src/package.js:708:39
  at Emitter.module.exports.Emitter.emit (/Applications/Atom.app/Contents/Resources/app/node_modules/event-kit/lib/emitter.js:82:11)
  at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/Applications/Atom.app/Contents/Resources/app/src/command-registry.js:224:20)
  at /Applications/Atom.app/Contents/Resources/app/src/command-registry.js:3:61
  at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:558:16)
  at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:396:22)
  at HTMLDocument.module.exports.WindowEventHandler.onKeydown (/Applications/Atom.app/Contents/Resources/app/src/window-event-handler.js:172:20)

```

### Commands

```
  4x -0:51.7 core:undo (atom-text-editor.editor.is-focused)
```

### Config

```json
{
  "core": {
    "excludeVcsIgnoredPaths": false,
    "disabledPackages": [
      "sublime-tabs",
      "linter-coffeelint"
    ],
    "audioBeep": false
  },
  "editor": {
    "fontSize": 12,
    "tabLength": 4,
    "invisibles": {}
  }
}
```

### Installed Packages

```coffee
# User
coffee-lint, v0.7.3
color-picker, v1.4.4
file-icons, v1.4.8
highlight-selected, v0.8.0
jsformat, v0.7.18
jshint, v1.2.0
language-marko, v0.4.8
linter, v0.10.0
pain-split, v1.2.0
project-manager, v1.14.1

# Dev
No dev packages
```

### LINKED ISSUE #5743: Failed to activate package - can we improve user notification ?
I am the developer of a package https://atom.io/packages/supercollider

Recently the package stopped working due to 6to5 raising an issue with unsafe use of eval in a dependency of a dependency of a dependency of atom-supercollider. Its functionality that is not even being used in atom-supercollider

1. The user isn't informed that the package failed to activate, has to look in console 
2. There is no report in the Packages pane for atom-supercollider
3. Developer (moi) who uses apm link does not experience this particular issue

For 1 I think that the nice red pop up panel should have been activated and allow the user to report the issue.

For 2 I suggest marking the package as "failed to activate" in some way. But maybe handling 1 is enough.

For 3: As a developer I was unaware that this was happening. I'm using the package with apm link so it doesn't get processed in the same way.  I had to unlink and then install it to see the error.

I can separate these issues if you want.

Error posted below. 

Its depd (which ironically is a package for reporting deprecations) which is in express which is used by supercollider.js (my npm module). Express isn't used by my atom package, so I can factor that functionality into a separate npm module.

```
 Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'".
/Users/crucial/github/atom-supercollider/node_modules/supercolliderjs/node_modules/express/node_modules/depd/index.js:414 wrapfunction
/Applications/Atom.app/Contents/Resources/app/src/package.js:241 Failed to activate package named 'atom-supercollider' EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'".

  at Function.wrapfunction (/Users/crucial/github/atom-supercollider/node_modules/supercolliderjs/node_modules/express/node_modules/depd/index.js:414:5)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/node_modules/supercolliderjs/node_modules/express/node_modules/send/index.js:156:47)
  at Module._compile (module.js:475:26)
  at Object.loadFile [as .js] (/Applications/Atom.app/Contents/Resources/app/src/6to5.js:138:21)
  at Module.load (module.js:370:32)
  at Function.Module._load (module.js:325:12)
  at Module.require (module.js:380:17)
  at require (module.js:399:17)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/node_modules/supercolliderjs/node_modules/express/lib/utils.js:16:12)
  at Module._compile (module.js:475:26)
  at Object.loadFile [as .js] (/Applications/Atom.app/Contents/Resources/app/src/6to5.js:138:21)
  at Module.load (module.js:370:32)
  at Function.Module._load (module.js:325:12)
  at Module.require (module.js:380:17)
  at require (module.js:399:17)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/node_modules/supercolliderjs/node_modules/express/lib/application.js:15:15)
  at Module._compile (module.js:475:26)
  at Object.loadFile [as .js] (/Applications/Atom.app/Contents/Resources/app/src/6to5.js:138:21)
  at Module.load (module.js:370:32)
  at Function.Module._load (module.js:325:12)
  at Module.require (module.js:380:17)
  at require (module.js:399:17)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/node_modules/supercolliderjs/node_modules/express/lib/express.js:7:13)
  at Module._compile (module.js:475:26)
  at Object.loadFile [as .js] (/Applications/Atom.app/Contents/Resources/app/src/6to5.js:138:21)
  at Module.load (module.js:370:32)
  at Function.Module._load (module.js:325:12)
  at Module.require (module.js:380:17)
  at require (module.js:399:17)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/node_modules/supercolliderjs/node_modules/express/index.js:2:18)
  at Module._compile (module.js:475:26)
  at Object.loadFile [as .js] (/Applications/Atom.app/Contents/Resources/app/src/6to5.js:138:21)
  at Module.load (module.js:370:32)
  at Function.Module._load (module.js:325:12)
  at Module.require (module.js:380:17)
  at require (module.js:399:17)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/node_modules/supercolliderjs/lib/nodejs/bridge.js:39:13)
  at Module._compile (module.js:475:26)
  at Object.loadFile [as .js] (/Applications/Atom.app/Contents/Resources/app/src/6to5.js:138:21)
  at Module.load (module.js:370:32)
  at Function.Module._load (module.js:325:12)
  at Module.require (module.js:380:17)
  at require (module.js:399:17)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/node_modules/supercolliderjs/index.js:6:18)
  at Module._compile (module.js:475:26)
  at Object.loadFile [as .js] (/Applications/Atom.app/Contents/Resources/app/src/6to5.js:138:21)
  at Module.load (module.js:370:32)
  at Function.Module._load (module.js:325:12)
  at Module.require (module.js:380:17)
  at require (module.js:399:17)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/lib/repl.coffee:6:19)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/lib/repl.coffee:1:1)
  at Module._compile (module.js:475:26)
  at Object.requireCoffeeScript (/Applications/Atom.app/Contents/Resources/app/node_modules/coffee-cash/lib/coffee-cash.js:67:19)
  at Module.load (module.js:370:32)
  at Function.Module._load (module.js:325:12)
  at Module.require (module.js:380:17)
  at require (module.js:399:17)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/lib/controller.coffee:2:8)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/lib/controller.coffee:1:1)
  at Module._compile (module.js:475:26)
  at Object.requireCoffeeScript (/Applications/Atom.app/Contents/Resources/app/node_modules/coffee-cash/lib/coffee-cash.js:67:19)
  at Module.load (module.js:370:32)
  at Function.Module._load (module.js:325:12)
  at Module.require (module.js:380:17)
  at require (module.js:399:17)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/lib/supercollider.coffee:1:14)
  at Object.<anonymous> (/Users/crucial/github/atom-supercollider/lib/supercollider.coffee:1:1)
  at Module._compile (module.js:475:26)
  at Object.requireCoffeeScript (/Applications/Atom.app/Contents/Resources/app/node_modules/coffee-cash/lib/coffee-cash.js:67:19)
  at Module.load (module.js:370:32)
  at Function.Module._load (module.js:325:12)
  at Module.require (module.js:380:17)
  at require (module.js:399:17)
  at Package.module.exports.Package.requireMainModule (/Applications/Atom.app/Contents/Resources/app/src/package.js:651:34)
  at Package.module.exports.Package.activateConfig (/Applications/Atom.app/Contents/Resources/app/src/package.js:251:12)
  at Package.module.exports.Package.activateNow (/Applications/Atom.app/Contents/Resources/app/src/package.js:232:14)
  at /Applications/Atom.app/Contents/Resources/app/src/package.js:710:29
  at Emitter.module.exports.Emitter.emit (/Applications/Atom.app/Contents/Resources/app/node_modules/event-kit/lib/emitter.js:82:11)
  at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/Applications/Atom.app/Contents/Resources/app/src/command-registry.js:224:20)
  at /Applications/Atom.app/Contents/Resources/app/src/command-registry.js:3:61
  at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:558:16)
  at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:396:22)
  at HTMLDocument.module.exports.WindowEventHandler.onKeydown (/Applications/Atom.app/Contents/Resources/app/src/window-event-handler.js:176:20)
```

### PR DESCRIPTION
**Pull request by kevinsawicki**
Calling `webkitMatchesSelector(selector)` on an element will throw an error. There were several places in core where these errors weren't being caught and handled.
- [x] Throw an error when `atom.commands.add` is called with an invalid selector
- [x] Throw an error when `atom.contextMenu.add` is called with an invalid selector
- [x] Show a notification when a package has invalid selectors in their activation commands
- [x] Show a notification when a package has invalid selectors in their context menu
- [x] Show a notification when a package has an invalid grammar
- [x] Show a notification when a package has invalid settings
- [x] Show a notification on all caught package load/activate errors

Closes https://github.com/atom/atom/issues/5098
Closes https://github.com/atom/atom/issues/5536
Closes #5743
Closes https://github.com/atom/command-palette/issues/30
Closes https://github.com/atom/command-palette/issues/32

The new notifications look like this when a package fails to load or activate:

screen shot 2015-03-03 at 6 13 01 pm

### CODE REVIEW

File: src/command-registry.coffee [hunk 1]
```diff
@@ -87,6 +88,7 @@ class CommandRegistry
       return disposable
 
     if typeof target is 'string'
+      validateSelector(target)
```
Comment (kevinsawicki) [inline src/command-registry.coffee]: @nathansobo you cool validating the selectors up front and throwing errors when they aren't valid? This way we don't need to `try/catch` around any `webkitMatchesSelector` calls

Comment (nathansobo) [inline src/command-registry.coffee]: Sounds very smart.

**Comment by nathansobo** [comment]
@kevinsawicki: It seems like `selector-parser.coffee` should really be named `selector-validator.coffee`, since that's all it does. Parsing to me implies it could return structural information.

**Comment by kevinsawicki** [comment]
> It seems like selector-parser.coffee should really be named selector-validator.coffee, since that's all it does. Parsing to me implies it could return structural information.

Sounds good

**Comment by Zireael07** [comment]
Thank you so much!!! This will make figuring out what went wrong with packages so much easier!

**Pull request merged**